gufe/base.py

```diff
@@ -0,0 +1,229 @@
+        return componentcls
+
+
+class _ABCGufeClassMeta(_GufeTokenizableMeta, abc.ABCMeta):
+    # required to make use of abc.ABC in classes that use _ComponentMeta metaclass
+    # see https://stackoverflow.com/questions/57349105/python-abc-inheritance-with-specified-metaclass 
+    ...
+        
+
+class GufeTokenizable(metaclass=_ABCGufeClassMeta):
+    """Base class for all tokenizeable gufe objects.
+
+    """
+    def __init__(self):
+
+        # since __hash__ isn't inheritable, monkey-patching is the best we can
+        # do besides defining it in every subclass
+        self.__class__.__hash__ = GufeTokenizable.__hash__
+
+    def __lt__(self, other):
+        return hash(self) < hash(other)
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return NotImplemented("Component comparisons must be of the same type")
+
+        return self.to_dict() == other.to_dict()
+
+    def __hash__(self):
+        return hash(self.key)
+
+    @property
+    def token(self):
+        if not hasattr(self, '_token') or self._token is None:
+            self._token = tokenize(self)
+        return self._token
+
+    @property
+    def key(self):
+        if not hasattr(self, '_key') or self._key is None:
+            prefix = type(self).__name__
+            self._key = GufeKey(f"{prefix}-{self.token}")
+
+        # add to registry if not already present
+        TOKENIZABLE_REGISTRY[self._key] = self
+
+        return self._key
+
+    @property
+    def defaults(self):
+        sig = inspect.signature(self.__init__)
+
+        defaults = {
+            param.name: param.default for param in sig.parameters.values()
+            if param.default is not inspect.Parameter.empty
+        }
+
+        return defaults
+
+    # TODO: rewrite the below three functions to be fully recursive like 
+    # _dictdecode_dependencies
+    def _keyencode_dependencies(self, d):
+        for key, value in d.items():
+            if isinstance(value, dict):
+                self._keyencode_dependencies(value)
+            elif isinstance(value, list):
+                for i, item in enumerate(value):
+                    if hasattr(item, '_gufe_tokenize'):
+                        d[key][i] = item.key
+            else:
+                if hasattr(value, '_gufe_tokenize'):
+                    d[key] = value.key
+
+        return d
+
+    @classmethod
+    def _keydecode_dependencies(cls, d):
+        for key, value in d.items():
+            if isinstance(value, dict):
+                cls._keydecode_dependencies(value)
+            elif isinstance(value, list):
+                for i, item in enumerate(value):
+                    if isinstance(item, str) and item in TOKENIZABLE_REGISTRY:
+                        d[key][i] = TOKENIZABLE_REGISTRY[item]
+            else:
+                if isinstance(value, str) and value in TOKENIZABLE_REGISTRY:
+                    d[key] = TOKENIZABLE_REGISTRY[value]
+
+        return d
```

Comment:
This seems to expect that objects that are being loaded are already in the registry. This will not be true when deserializing.

Reply:
Actually, we can require storage to populate `TOKENIZABLE_REGISTRY` with everything it contains when it opens. Then this would work. The assumption is that there are relatively few objects to be stored there -- although it could get into the thousands, so I'm a little concerned about performance on opening storage in that case (especially if each object is a separate call to a cloud service.)

The alternative is that we need to provide a function to load something from storage as part of the signature here. Even with that, it's a bit tricky.

Reply:
Do the methods we have now give a storage implementation what it needs to identify (recursively) the dependents and pull them out before fully instantiating the requested object?

Reply:
That would still be "on-demand", instead of requiring everything to be in-memory all at once that has ever been stored.